protocol/implementations/js/src/builders.ts

```diff
@@ -0,0 +1,36 @@
+import { MessageTypes, TbDEXMessage } from './types.js'
+import {ulid} from 'ulidx'
+
+// maybe extract this somewhere?
+const ID_NAMESPACE = 'tbdex'
+
+function createId(messageType: string): string {
+  return `${ID_NAMESPACE}:${messageType}:${ulid()}`
+}
+
+export function createMessage<Type extends keyof MessageTypes>(to: string, from: string, body: MessageTypes[Type], messageType: Type,): TbDEXMessage<Type> {
+  // TODO could validate that type is Offering or RFQ, since they start chains or are one off types
+  const id = createId(messageType)
+  return {
+    id          : id,
+    contextId   : id,
```

Comment:
At what point is `contextId` different from `id`? Or will it always be different and you're just stubbing this in temporarily right now?

Reply:
Just showcasing different syntax approach, one is not necessarily better than the other. I like exploring different syntax approaches for the intent of improving readability. Of course, readability, like beauty is in the eye of the beholder 😉 

```typescript
interface ICreateMessage {
  <Type extends keyof MessageTypes>(to: string, from: string, type: Type, body: MessageTypes[Type]): TbDEXMessage<Type>
}
export const createMessage: ICreateMessage = (to, from, type, body) => {
  const id = createId(type)
  return {
    id,
    contextId   : id,
    createdTime : new Date().toISOString(),
    to,
    from,
    type,
    body,
  }
}
```

Reply:
btw if it weren't for the matter that `id` and `contextId` are the same thing, I would push it further to this... 

```typescript
export const createMessage: ICreateMessage = (to, from, type, body) => ({
  id: createId(type),
  contextId   : 'something else???',
  createdTime : new Date().toISOString(),
  to,
  from,
  type,
  body,
})
```

To me, this is very explicit in the intent... which is something like... _yeah the developer still have to pass in a majority of the properties, so it's not much different than just having them create the object on their own, but we're abstracting away the id & timestamping, which some devs may find verbose_

Reply:
> At what point is contextId different from id?

It is always the same in this method! in the below function, where you pass in the last message, `contextId` will be different (pulled from the existing thread of messages).

Reply:
@KendallWeihe bc I'm a TS newb, what do you think is the benefit of using `interface ICreateMessage`+`const createMessage: ICreateMessage` over a function type? Is this pattern still used when the interface only has one implementation?

Reply:
`const` for the function (rather than `function`) is purely preferential... I like it because it reinforces the concept that everything in js is an object

> Is this pattern still used when the interface only has one implementation?

There is no pattern here, that I'm aware of. Again, purely subjective. I actually usually prefer inline typing, but in this case, there are so many types... idk to me it's easier to read by splitting it out

So yeah, beauty is in the eye of the beholder